mindscope_utilities/visual_behavior_ophys/data_formatting.py

```diff
@@ -376,3 +376,208 @@ def get_stimulus_response_df(ophys_experiment,
+
+        # get necessary tables
+        stimulus_presentations = ophys_experiment.stimulus_presentations.copy()
+        rewards = ophys_experiment.rewards.copy()
+
+        # add rewards to stim presentations
+        stimulus_presentations = add_rewards_to_stimulus_presentations(stimulus_presentations, rewards)
+    '''
+
+    reward_times = rewards['timestamps'].values
+    rewards_each_stim = stimulus_presentations.apply(
+        lambda row: reward_times[
+            ((reward_times > row["start_time"] +
+              range_relative_to_stimulus_start[0]) & (
+                reward_times < row["start_time"] +
+                range_relative_to_stimulus_start[1]))],
+        axis=1,
+    )
+    stimulus_presentations["rewards"] = rewards_each_stim
+    return stimulus_presentations
+
+
+def add_licks_to_stimulus_presentations(
+    stimulus_presentations,
+    licks,
+    range_relative_to_stimulus_start=[
+        0,
+        0.75]):
+    '''
+    Append a column to stimulus_presentations which contains the timestamps of licks that occur
+    in a range relative to the onset of the stimulus.
+
+    Args:
+        stimulus_presentations (pd.DataFrame): dataframe of stimulus presentations.
+            Must contain: 'start_time'
+        licks (pd.DataFrame): lick dataframe. Must contain 'timestamps'
+        range_relative_to_stimulus_start (list with 2 elements): start and end of the range
+            relative to the start of each stimulus to average the running speed.
+    Returns:
+        stimulus_presentations with a new column called "licks" that contains
+        lick times that fell within the window relative to each stim time
+
+
+    Example:
+        # get visual behavior cache
+        from allensdk.brain_observatory.behavior.behavior_project_cache import VisualBehaviorOphysProjectCache as bpc
+        cache_dir = r'\\allen\programs\braintv\workgroups\nc-ophys\visual_behavior\platform_paper_cache'
+        cache = bpc.from_s3_cache(cache_dir=cache_dir)
+
+        # load data for one experiment
+        ophys_experiment = cache.get_behavior_ophys_experiment(experiment_id)
+
+        # get necessary tables
+        stimulus_presentations = ophys_experiment.stimulus_presentations.copy()
+        licks = ophys_experiment.licks.copy()
+
+        # add licks to stim presentations
+        stimulus_presentations = add_licks_to_stimulus_presentations(stimulus_presentations, licks)
+    '''
+
+    lick_times = licks['timestamps'].values
+    licks_each_stim = stimulus_presentations.apply(
+        lambda row: lick_times[
+            ((lick_times > row["start_time"] +
+              range_relative_to_stimulus_start[0]) & (
+                lick_times < row["start_time"] +
+                range_relative_to_stimulus_start[1]))],
+        axis=1,
+    )
+    stimulus_presentations["licks"] = licks_each_stim
+    return stimulus_presentations
+
+
+def get_trace_average(trace, timestamps, start_time, stop_time):
```

Comment:
ah, i was not aware of 'compute_means'. Since these functions are on the backend of the functions we will be asking students to use, it shouldnt be too disruptive if we change the way it gets means behind the scenes. Let's remember to fix this once the notebooks are done